### ISSUE TITLE
XML Markup Scheme for Marian Decoder

### ISSUE THREAD

**Issue by phikoehn** [marian-nmt/marian#157]
I might be interested in taking another stab at a XML markup scheme with specified translations as in Moses:

The house is <x translation="klein"> small </a> .

This does require the decoding algorithm to have access to the attention states for each output word. This sounds rather tricky to me, since the decoder is model-agnostic. Any suggestions?

**Comment by emjotde** [comment]
I guess starting in models/encdec.h, if you look at this line:

https://github.com/marian-nmt/marian-dev/blob/master/src/models/encdec.h#L355

you see how to get access to the attention weights of a decoder, currently that only works for RNN-based decoder. I haven't unified this yet. You would need to push it through to scorer in

https://github.com/marian-nmt/marian-dev/blob/master/src/translator/scorers.h#L59

as they hold the EncDec object during translation. Then you are already at the level of the translator.

**Comment by emjotde** [comment]
Adding actual XML to our very single purpose input/output formats may actually be the more challenging part :)

**Comment by lkfo415579** [comment]
@emjotde 1. I had tried the case "only overwrite the target token".
I chosed using attention matrix to do replacement. ( En : I played <unk>FOOTBALL</unk>, Fr : j'ai joué au UNK. , After Replacement : j'ai joué au FOOTBALL, Attention : 2-3) but **it won't able** to generate UNK or having a correct attention matrix after translation in sometimes. (complicated sentence will fail as well)
2. force an embedding while decoding is a great idea but it may fail during OOV.

**Comment by mtresearcher** [comment]
Reviving this, was this handled at some point? 
If not, I can add this functionality (overwriting the target token) in the PR here: https://github.com/ebay-hlt/marian-dev/pull/1

**Comment by phikoehn** [comment]
I started a branch (xml) in marian-nmt / marian-dev. I am currently working on it and hope to get somewhere in the next 1-2 weeks (implementing a grid search approach).

**Comment by mtresearcher** [comment]
@phikoehn I have already implemented XML handling in the above PR (minute effort is needed for alternate translations). If it makes it any easy for you I can issue a PR to marian-dev and you can build on top of it.

**Comment by phikoehn** [comment]
Looking over the code, this introduces a word replacement feature, where the decoder is expected to produce special output tokens (starting with $) that are then replaced during output printing. This works fine for named entities (this seems to be the intention), but not for words who influence word choices around it. I am planning to modify the search algorithm (using "grid search") to also allow multi-word XML tags.

**Comment by mtresearcher** [comment]
Yeah. In eBay we have placeholders starting with $ and the idea for now was just to replace those with the entities.

**Comment by phikoehn** [comment]
I completed the XML constraint decoding. The code currently checked into XML branch is fully functional, also with batched decoding but could use some stress testing (I am planning to do that) and is currently spewing out excessive amounts of debugging information. --- I would appreciate help with two things to make it ready to be merged into the main branch. 

(1) The memory management for XML-related data structures does not use smart pointers and is currently leaking memory. Can someone who understands the Marian pointers take a look at that? I can give detailed guidance on where objects are allocated. 

(2) The code is not merged with the latest version of Marian, partly because I have a problem with compiling the latest version of Marian with the ancient libraries that are on my systems.

**Comment by snukky** [comment]
I'm happy to help.
